File: src/days/day_02.rb
class Day

  def call input
    return "#{a(input)}, #{b(input)}"
  end

  def a input
    result = input.lines.reduce({horizontal: 0, depth: 0}) do |acc, line|
      instruction, amount = line.split(" ")
      amount = amount.to_i

      case instruction
      when "forward"
        acc[:horizontal] += amount
      when "down"
        acc[:depth] += amount
      when "up"
        acc[:depth] -= amount
      else
        puts "Did not know what to do with instruction: #{}"
      end

      acc
    end

    result[:horizontal] * result[:depth]
  end

  def b input
    result = input.lines.reduce({aim: 0, horizontal: 0, depth: 0}) do |acc, line|
      instruction, amount = line.split(" ")
      amount = amount.to_i

      case instruction
      when "forward"
        acc[:horizontal] += amount
        acc[:depth] += amount * acc[:aim]
      when "down"
        acc[:aim] += amount
      when "up"
        acc[:aim] -= amount
      else
        puts "Did not know what to do with instruction: #{}"
      end

      acc
    end

    result[:horizontal] * result[:depth]
  end
end

File: src/days/day_02.rs
struct Instruction {
    direction: String,
    amount: i32,
}


pub fn call(puzzle_input: String) -> String {
    let lines = puzzle_input.lines();

    let instructions: Vec<Instruction> = lines.map(|line| {
        let mut parts = line.split(" ");

        let direction = parts.next().unwrap().to_string();
        let amount = parts.next().unwrap().parse::<i32>().unwrap();

        Instruction {
            direction,
            amount
        }
    })
    .collect();

    format!("{}, {}", a(&instructions), b(&instructions))
}

fn a(instructions: &Vec<Instruction>) -> i32 {
    let result = instructions.iter().fold((0,0), |mut acc, instruction| {
        match instruction.direction.as_str() {
            "forward" => acc.0 += instruction.amount,
            "down" => acc.1 += instruction.amount,
            "up" => acc.1 -= instruction.amount,
            _ => ()
        };

        acc
    });

    result.0 * result.1
}

fn b(instructions: &Vec<Instruction>) -> i32 {
    let result = instructions.iter().fold((0,0,0), |mut acc, instruction| {
        match instruction.direction.as_str() {
            "down" => acc.2 += instruction.amount,
            "up" => acc.2 -= instruction.amount,
            "forward" => {
                acc.0 += instruction.amount;
                acc.1 += instruction.amount * acc.2
            }
            _ => ()
        };

        acc
    });

    result.0 * result.1
}
